### ISSUE TITLE
Improvement: percentage rating

### ISSUE THREAD

**Issue by stdoma** [gbateson/moodle-qtype_essayautograde#77]
Hello!
In some cases, the percentage valuation in 1% steps can be a hindrance.
I wanted to use the plugin as an extension of short answers.
I have a list of words from which x should be named.

If I want three words from the list, I have to give an answer 33 or 34% of the points. However, this does not give me an even 100% if all answers should or must have the same weighting.

In some cases it would be better to use percentage scoring, e.g. for normal short answers.
Is it possible to set up intermediate levels or adjust the information yourself?

**Comment by gbateson** [comment]
Hi Stephan,
thanks for your message. I'm not sure what you mean by the phrase, "I have a list of words from which x should be named". Could you elaborate on what "x" is, and what you mean by "named"?

Also, I'm not sure what you mean by "percentage scoring, e.g. for normal short answers". Do you mean that the teacher inputs the percentage into a text box? It's still impossible to input exactly one third of 100%. Would you input 33.3333333%

Or perhaps you wish to award "100% divided by the number of target words" for each target word?

Or perhaps oyu mean something else. Sorry if I have misunderstood.

**Comment by stdoma** [comment]
Hi, thanks for asking.
I had thought that I could use the plugin for a teacher's request.

In some of his questions, he wants students to name words. There is a list of words with x elements. From these, the students only need to name y.
If, for example, 3 words are to be named, each answer option in the list should get 33.33%.

In the plugin there are integer values for the percentages.
In other question types, such as the short answer, there is this selection list for the percentages: 100, 90, 83.33, 80, 75, 70, 66.66, 60, 50, 40, 30, 20, 16.66, 14.285, 12.5, 11,11, 10, 5
(From my point of view, it is not necessary to be able to assign every value between 1 and 100 to the single answers).

If you only have integer values available, you have to adjust the number of answers.
I would find it more flexible if you had more leeway with the percentage values.

**Comment by gbateson** [comment]
So having a list of percentages that included 83.33, 66.66, 16.66, 14.285 and 12.5 would suit your needs? I didn't know about "14.285". That seems to be 100% / 7. Those fractional number seems a little overly mathematical. I wonder if we could rewrite them more clearly for language teachers as
"1/8 of the question score" up to "7/8 of the question score"
"1/7 of the question score" up to "6/7 of the question score"
"1/6 of the question score" up to "5/6 of the question score"
"1/3 of the question score" up to "2/3 of the question score"

That's getting quite long-winded though, isn't it.

More generally, we could phrase it as "xx% equally shared between yy number of answers"
So for your teacher, it would be "100% shared equally between 3 answers".

This would allow a slightly more complex scoring where 1 target phrase is worth 20%, and the remaining 80% is shared equally between three other phrases.

if "Target-phrase-1" is used, award [20%] of the question grade, shared equally between [1] answers.
if "Target-phrase-2" is used, award [80%] of the question grade, shared equally between [3] answers.
if "Target-phrase-3" is used, award [80%] of the question grade, shared equally between [3] answers.
if "Target-phrase-4" is used, award [80%] of the question grade, shared equally between [3] answers.

Or perhaps the following is clear enough:

if "Target-phrase-1" is used, award [20%] [ exactly ] of the question grade.
if "Target-phrase-2" is used, award [80%] [divided by 3] of the question grade.
if "Target-phrase-3" is used, award [80%] [divided by 3] of the question grade.
if "Target-phrase-4" is used, award [80%] [divided by 3] of the question grade.

For your teacher, this would appear as

if "Target-phrase-1" is used, award [100%] [divided by 3] of the question grade.
if "Target-phrase-2" is used, award [100%] [divided by 3] of the question grade.
if "Target-phrase-3" is used, award [100%] [divided by 3] of the question grade.

What do you think of this approach?

**Comment by stdoma** [comment]
Yes, that would be exactly what I was looking for and what other people might need ;)
It would also be a big step forward for the possibilities of ratings.

I think a short spelling is enough.
Teachers who use tests in this way are usually a little more advanced in their Moodle knowledge.

**Comment by gbateson** [comment]
We might have to be a little ingenious to store this information in the "question_answers" table without adding any more fields to the database. The moment the percentage for a target word is stored in the "feedback_format" of the question_answers table. Looks like a tricky problem. Please let me think about that for a while ...

**Comment by gbateson** [comment]
I've been experimenting with the input form for these new settings. This is the current state.
<img width="932" alt="Screenshot 2024-03-19 at 9 52 41" src="https://github.com/gbateson/moodle-qtype_essayautograde/assets/799312/2927339a-18bc-4ca0-a52a-37e72ebf87e0">

**Comment by gbateson** [comment]
I have added 4 new language strings that fit before, between and after the input elements.
i.e. {text-1} [target-phrase] {text-2} [percent] {text-3} [divisor] {text-4}

in the English language pack, these strings will be defined as:
text-1 = If
text-2 = is used, award
text-3 = (empty)
text-4 = of the question grade.

Do you speak any other languages? Can you imagine what these strings would be in other language that you are familiar with?

**Comment by stdoma** [comment]
Hello!
That went faster than expected ;)
With the solution, existing questions would not need to be adapted. That looks good.


I don't know enough about UX aspects:
- As an idea, you could make the "divided by" combo box invisible first and make it visible with a checkbox ("use partial percentages"). However, this would give you even more graphical elements in a question field.
- Another variant would be to place the "divided by" combo box in front of the percentage combo box:
-- if [TARGET] is used, award [divisor] from [percent] of the question grade.
-- if "test" is used, award "1/1" from "80%" of the question grade.
-- if "test" is used, award "1/3" from "80%" of the question grade.


I could help with the translation for German language packs.
I just looked into the language pack and realized that not all elements / strings have a German translation.
